### ISSUE TITLE
[Bug]: Issue with The Default Axis Limits of Errorbar Plots

### ISSUE THREAD

**Issue by wei-lingfeng** [matplotlib/matplotlib#25604]
### Bug summary

When there are nans in the pandas.DataFrame, the errorbar function fails to automatically adjust axis limits to reflect the valid data only. The axis limit is subject to the extremist values in the dataframe, but that data point is not necessarily plotted out. See the example below.

### Code for reproduction

```python
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

test = pd.DataFrame({
    'x': np.array([0, 1, 2, np.nan, 10]),
    'y': np.array([4, 5, 3, 10, np.nan]),
    'xerr': np.array([.4, .3, .5, np.nan, .2]),
    'yerr': np.array([.3, .6, .5, .2, np.nan])
})
plt.errorbar(test.x, test.y, test.xerr, test.yerr)
plt.show()
```

### Actual outcome

![image](https://user-images.githubusercontent.com/34182375/229433128-6683939b-b92e-43ec-92b8-d0395d4e7891.png)

### Expected outcome

![image](https://user-images.githubusercontent.com/34182375/229431701-c99ac95f-371f-4e8d-9a8f-0749c41203f0.png)

### Additional information

Matplotlib 3.5.2 was able to adjust the axis limits according to all non-nan values automatically.

### Matplotlib Version

3.7.1

### Python version

3.11.2

### Installation

None

**Comment by rcomer** [comment]
Thanks for the report @wei-lingfeng.  I have reproduced the above behaviour with our `main` development branch.  However if I plot the arrays directly, I get

```python
import numpy as np
import matplotlib.pyplot as plt

x = np.array([0, 1, 2, np.nan, 10])
y = np.array([4, 5, 3, 10, np.nan])
xerr = np.array([.4, .3, .5, np.nan, .2])
yerr = np.array([.3, .6, .5, .2, np.nan])

plt.errorbar(x, y, xerr, yerr)

plt.show()
```
![test](https://user-images.githubusercontent.com/10599679/229440021-a44fe2a2-9e4f-4c1d-9b3f-4ce33b77b35a.png)

So the problem appears to be in how pandas and matplotlib are interacting.

**Comment by rcomer** [comment]
I also tried with Matplotlib 3.5.2 and Pandas 1.5.3 (latest), and got this:

![test](https://user-images.githubusercontent.com/10599679/229444784-8f241093-b00a-4cc8-b3c1-a83e0ece4759.png)

**Comment by oscargus** [comment]
As a workaround for now, (if you still want to your data stored as a pandas dataframe) you can use
```
C = plt.errorbar(test.x.values, test.y.values, test.xerr.values, test.yerr.values)
```
This is roughly what the code should be doing, but there's been some changes with respect to NaN-handling recently.

**Comment by ksunden** [comment]
Note that what @oscargus was referring to was not the float vs int, but rather handling inside of Matplotlib itself which is done with the explicit goal of preserving unit information when it is provided. (if you do not hand `errorbar` an `ndarray` (specifically) it will cast, using `np.asndarray(x, dtype=object)` since `test.x` is a pandas Series, it hits that code path, regardless of what the pandas dtype is.

**Comment by dstansby** [comment]
On current Matplotlib main and pandas 2.1.4 I get the below figure, which I think is 'correct', so maybe we can close this as fixed?
<img width="610" alt="Screenshot 2024-01-14 at 18 43 32" src="https://github.com/matplotlib/matplotlib/assets/6197628/f66ae8c6-1e5a-4666-b39a-13de4ad81947">

**Comment by QuLogic** [comment]
I can also confirm this works on Matplotlib 3.8.2 and Pandas 1.5.3, so I believe we must have fixed in 3.8+ some time.
